### ISSUE TITLE
Preferences have local paths

### ISSUE THREAD

**Issue by jjatria** [rolandomunoz/plugin_spanish_voiced_stops_v2.0.0#4]
The [`.preferences/preferences.txt`](https://github.com/rolandomunoz/plugin_spanish_voiced_stops_v2.0.0/blob/5b4edb8adcfe1842a1311a226d50973b00b4a74b/.preferences/preferences.txt) file has local directories for both the `textGrid_folder_dir` and `sound_folder_dir` variables. It is extremely unlikely that these will exist in other user's computers. Please try to make it so the plugin will work regardless of where it is run.

(Also: why use both "folder" _and_ "dir" in these variable names?).

Incidentally, you might want to take a look at the (acceptedly under-documented) [`config` plugin](http://cpran.net/plugins/config/) to read a config file into a local variable.

**Comment by rolandomunoz** [comment]
In this plugin, local variables ([`.preferences/preferences.txt`](https://github.com/rolandomunoz/plugin_spanish_voiced_stops_v2.0.0/blob/5b4edb8adcfe1842a1311a226d50973b00b4a74b/.preferences/preferences.txt)) can be changed by the user interface. Users can do it by using:  ` Preferences> Set > Set working directory` command in the plugin menu. This is supported by the `@lib_preference.preferences_set_field_content_from: preference_file$, "field_name", "field_content"` procedure. This  allows the user to modify the [`.preferences/preferences.txt`](https://github.com/rolandomunoz/plugin_spanish_voiced_stops_v2.0.0/blob/5b4edb8adcfe1842a1311a226d50973b00b4a74b/.preferences/preferences.txt) file.  <br \>
The config.proc procedures seems to have a clearer syntax. I'll take a look at it.  <br \>
Thank you.

**Comment by jjatria** [comment]
Yeah, I noticed you had that. And of course, you don't _have_ to use that plugin. I just find it easier. :)

Personally, if I wanted to give the user an interface to modify fields, I think I'd make a `pause` with all* the fields and their values, which would serve as a getter and a setter. Have you thought about that?

**Comment by rolandomunoz** [comment]
That's exactly how this works. The preferences fields are read by a `get method`, then the information is put on a `pause` window. Once the user decides to continue, the data is rewritten in the preferences file. I use a `write method`for this purpose.
In the next realises, I will delete the `Preference > Get ` menu, and only include the `Set` commands, which also show the preference fields.

I have decided to generalize this plug-in, so it can be applied to other cases out of the scope of the project that it was created for. It will take a time to complete this task. I hope I can tell you when things are done. Your advices are very helpful Thank you

By the other hand, I have written a simple plug-in which is helpful for linguistic segmentation of data. I use it when I transcribe spoken languages from Peru. I would like to share this plug-in with other Praat users through CPrAN. May you take a look at it? Here is the website: https://github.com/rolandomunoz/plugin_tokenizer At this moment I'm focusing in writing the documentation. Thank you in advance.

**Comment by jjatria** [comment]
Looks good! I noticed that there are a couple of procedures you've put together that already exist in CPrAN. In particular, you might want to take a look at the `strutils` plugin, which does a lot of what your `strings_object.proc` file does.

The [`array.proc` procedures](https://gitlab.com/cpran/plugin_strutils/blob/master/procedures/array.proc) provide an array-like interface for String objects, including `@split` (which is equivalent to your `strings.create_Strings_from`), and `@join` (which is equivalent to `strings.get_string_var`); and the [`Sort Strings (generic)...` script](https://gitlab.com/cpran/plugin_strutils/blob/master/scripts/sort_strings_generic.praat) can do something similar to what you do in `strings.sort_by_nchars`.

There's a lot of good stuff already written in those plugins, which have been designed to be used by other projects. Take a look! And if you find that they are not suitable, pull requests and suggestions are more than welcome.

I'll try to make more comments in that repository as they come up.

> That's exactly how this works. 

I thought you asked the user for a specific field. What I meant was that you could give the user a form with _all_ the variables in the configuration file, and let them edit them there, instead of asking the user for a specific variable to set / get.

**Comment by rolandomunoz** [comment]
Once again, thank you for your comments. I was taking a look to the  `strutls`  methods and other procedures a days ago. I'm starting to get better picture of the importance of working in general procedures. I was thinking that it will take me some time to accustom to the new features, but hope I can do it soon. Please, where are those projects ? In GitHub?

My little experience in programming comes from Python, there we have a set of core libraries that are available for all python users, as in R, C++, or Perl (I guess). So when I use a library, it comes by default with Python. In the other hand, at this moment, this is not possible for Praat (that the .proc comes predefined). This leads me to ask me how could I include all those procedures files from CPRaN to my own plug-ins. As they don't come by default with Praat, I need to download them using CPrAN or from Git repositories. Once this is done, I could directly include these files inside my own script folder, or I could  make reference to them by using a kind of inter-plugin path reference. In this last way, I could conserve both plug-in folders. I'm not clear about this point. How could I integrate procedures from CPRaN in my own plugins?

> I thought you asked the user for a specific field. What I meant was that you could give the user a form with all the variables in the configuration file, and let them edit them there, instead of asking the user for a specific variable to set / get.

In fact, `set` and `get` are internal process, so users doesn't deal with those procedures. Instead, users have a friendly interface, a pause window. It asks them, for example, for the directory where something is located or a tier name. After that, the information is taken from the pause windows and written in the preferences files by the `set` method. I use `get` to extract the last thing that were stored in preferences and show it in the pause windows.

**Comment by jjatria** [comment]
> where are those projects ? In GitHub?

The projects are all hosted on [GitLab](https://gitlab.com) (a free alternative to Github), and organised under [the `cpran` group](https://gitlab.com/cpran). But an effort at providing a web interface to their documentation is [on the CPrAN website](http://cpran.net).

> This leads me to ask me how could I include all those procedures files from CPRaN to my own plug-ins. As they don't come by default with Praat, I need to download them using CPrAN or from Git repositories.

[CPrAN plugins are just Praat plugins with some added sugar](http://cpran.net/docs/plugins/), so all of them can be installed manually just like you would with any other plugin: by downloading a local copy of the `plugin_foo` directory to your preferences directory. Each plugin page on the CPrAN website (eg. [here's the one for `strutils`](http://cpran.net/plugins/strutils)) links to an archive with the latest contents of that plugin, that you can download and install manually. In the case of `strutils`, [that's here](https://gitlab.com/cpran/plugin_strutils/repository/archive.zip), for example.

Installing these manually is a pain, though, because you have plugins that depend on other plugins, so there is also a specific order in which they need to be installed if you want each step to work.

To make all of this as easy as I can, I've made a CPrAN client, which takes care of finding the plugins in the CPrAN directory, installing them and their dependencies in the right order, testing them to make sure they work, and upgrade them whenever there are new releases. You can find [information on how to install the client](http://cpran.net/clients/cpran/#installation) on the CPrAN website.

After installing the client, you should run `cpran init`, which downloads the most basic bundle of plugins, and `cpran update` to get a refreshed list of releases from the server. Then you can use it normally.

Once you've installed the dependencies, you can use them from your own scripts by including the procedure files, or by running the scripts. This is an important distinction (which is also outlined in [the CPrAN plugin page](http://cpran.net/docs/plugins/)):

* When you include a procedure, the include path cannot be interpolated, so it needs to be hardcoded. You can use relative paths, but all those include paths will be relative _to the topmost script which is doing the including_. 
  To solve this, all the plugins on CPrAN use relative paths from the preferences directory, which means that they will only work if included from scripts that are themselves in plugins. In addition, all scripts who do the including need to be the same number of levels under the preferences directory, which means that they need to be in eg. `/home/foo/plugin_bar/scripts`. 
  That's the structure all the existing plugins use: a root directory with a `scripts`, and `procedures` directories, and all their scripts and procedures directly under these directories.

* Alternatively, scripts can be executed from anywhere, because the command used to execute them from within a script takes a path as a string, which can be interpolated. So you can specify the location of the script to run from wherever your current script is, and whatever files are included b the script you are executing will be relative _to that script_ (which will work as long as it's two levels under the preferences directory).

You can take a look at the code in the existing plugins, which all include procedures and execute scripts left and right.
